From d39a37a6057d4f8205895699b789a2965d2f7f56 Mon Sep 17 00:00:00 2001
From: Hongxu Jia <hongxu.jia@windriver.com>
Date: Fri, 28 Feb 2025 15:12:41 +0800
Subject: [PATCH] Add support for posix_devctl()

It is defined in POSIX 1003.26 and required by the FACE Technical
Standard.

POSIX 1003.26 is available from here:
https://standards.ieee.org/findstds/standard/1003.26-2003.html

Implement FACE 3.0 specific details, as documented in Section 3.2.1.2, GPSS
POSIX numbers 10 and 11, see:
http://oaevents.goprecise.com/wp-content/uploads/2017/11/FACE-Technical-Standard-Edition-3.0.pdf

Upstream-Status: Pending

Rebase to glibc 2.41

Signed-off-by: Hongxu Jia <hongxu.jia@windriver.com>
---
 bits/confname.h                               |   2 +
 posix/Makefile                                |   2 +
 posix/Versions                                |   5 +
 posix/devctl.h                                |  51 ++++++++
 posix/getconf.c                               |   2 +
 posix/posix_devctl.c                          |  31 +++++
 posix/sysconf.c                               |   7 ++
 posix/unistd.h                                |   4 +
 sysdeps/mach/hurd/i386/libc.abilist           |   1 +
 sysdeps/unix/sysv/linux/aarch64/libc.abilist  |   1 +
 sysdeps/unix/sysv/linux/alpha/libc.abilist    |   1 +
 sysdeps/unix/sysv/linux/arm/be/libc.abilist   |   1 +
 sysdeps/unix/sysv/linux/arm/le/libc.abilist   |   1 +
 sysdeps/unix/sysv/linux/csky/libc.abilist     |   1 +
 sysdeps/unix/sysv/linux/hppa/libc.abilist     |   1 +
 sysdeps/unix/sysv/linux/i386/libc.abilist     |   1 +
 .../sysv/linux/m68k/coldfire/libc.abilist     |   1 +
 .../unix/sysv/linux/m68k/m680x0/libc.abilist  |   1 +
 .../sysv/linux/microblaze/be/libc.abilist     |   1 +
 .../sysv/linux/microblaze/le/libc.abilist     |   1 +
 .../sysv/linux/mips/mips32/fpu/libc.abilist   |   1 +
 .../sysv/linux/mips/mips32/nofpu/libc.abilist |   1 +
 .../sysv/linux/mips/mips64/n32/libc.abilist   |   1 +
 .../sysv/linux/mips/mips64/n64/libc.abilist   |   1 +
 sysdeps/unix/sysv/linux/posix_devctl.c        | 116 ++++++++++++++++++
 .../linux/powerpc/powerpc32/fpu/libc.abilist  |   1 +
 .../powerpc/powerpc32/nofpu/libc.abilist      |   1 +
 .../linux/powerpc/powerpc64/be/libc.abilist   |   1 +
 .../linux/powerpc/powerpc64/le/libc.abilist   |   1 +
 .../unix/sysv/linux/riscv/rv64/libc.abilist   |   1 +
 .../unix/sysv/linux/s390/s390-32/libc.abilist |   1 +
 .../unix/sysv/linux/s390/s390-64/libc.abilist |   1 +
 sysdeps/unix/sysv/linux/sh/be/libc.abilist    |   1 +
 sysdeps/unix/sysv/linux/sh/le/libc.abilist    |   1 +
 .../sysv/linux/sparc/sparc32/libc.abilist     |   1 +
 .../sysv/linux/sparc/sparc64/libc.abilist     |   1 +
 .../unix/sysv/linux/x86_64/64/libc.abilist    |   1 +
 .../unix/sysv/linux/x86_64/x32/libc.abilist   |   1 +
 38 files changed, 249 insertions(+)
 create mode 100644 posix/devctl.h
 create mode 100644 posix/posix_devctl.c
 create mode 100644 sysdeps/unix/sysv/linux/posix_devctl.c

diff --git a/bits/confname.h b/bits/confname.h
index 2eaf8a2e..942286b8 100644
--- a/bits/confname.h
+++ b/bits/confname.h
@@ -492,6 +492,8 @@ enum
 #define _SC_LEVEL4_CACHE_ASSOC		_SC_LEVEL4_CACHE_ASSOC
     _SC_LEVEL4_CACHE_LINESIZE,
 #define _SC_LEVEL4_CACHE_LINESIZE	_SC_LEVEL4_CACHE_LINESIZE
+    _SC_POSIX_26_VERSION,
+#define _SC_POSIX_26_VERSION		_SC_POSIX_26_VERSION
     /* Leave room here, maybe we need a few more cache levels some day.  */
 
     _SC_IPV6 = _SC_LEVEL1_ICACHE_SIZE + 50,
diff --git a/posix/Makefile b/posix/Makefile
index a650abf5..66939438 100644
--- a/posix/Makefile
+++ b/posix/Makefile
@@ -51,6 +51,7 @@ headers := \
   bits/waitstatus.h \
   cpio.h \
   fnmatch.h \
+  devctl.h \
   getopt.h \
   glob.h \
   re_comp.h \
@@ -116,6 +117,7 @@ routines := \
   pathconf \
   pause \
   posix_madvise \
+  posix_devctl \
   pread \
   pread64 \
   pwrite \
diff --git a/posix/Versions b/posix/Versions
index 0624d24b..4b0ac410 100644
--- a/posix/Versions
+++ b/posix/Versions
@@ -159,6 +159,11 @@ libc {
   GLIBC_2.35 {
     posix_spawn_file_actions_addtcsetpgrp_np;
   }
+
+  GLIBC_2.28 {
+    posix_devctl;
+  }
+
   GLIBC_PRIVATE {
     __libc_fork; __libc_pread; __libc_pwrite;
     __nanosleep_nocancel; __pause_nocancel;
diff --git a/posix/devctl.h b/posix/devctl.h
new file mode 100644
index 00000000..1010035a
--- /dev/null
+++ b/posix/devctl.h
@@ -0,0 +1,51 @@
+/* Definitions for POSIX posix_devctl interface.
+   Copyright (C) 2014-2019 Wind River Systems, Inc.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <http://www.gnu.org/licenses/>. */
+
+#ifndef _DEVCTL_H
+#define _DEVCTL_H
+
+#include <unistd.h>
+/*
+ * posix_devctl() is defined by POSIX 1003.26-2003. It requires
+ * POSIX_26_C_SOURCE to be defined in order for the function to
+ * be available.
+ */
+#if _POSIX_26_C_SOURCE == 200312L
+
+#include <sys/ioctl.h>
+#include <sys/types.h>
+
+/* http://oaevents.goprecise.com/wp-content/uploads/2017/11/FACE-Technical-Standard-Edition-3.0.pdf
+ * Implement FACE 3.0 specific details, as documented in Section 3.2.1.2, GPSS
+ * POSIX numbers 10 and 11, see:
+ * An OSS UoC shall define:
+ * a. A macro in <sys/ioctl.h> named FIONBIO to configure sockets for non-blocking I/O
+ * b. A macro in <devctl.h> named SOCKCLOSE to close an open socket
+ * Use `_IOR' to define SOCKCLOSE that avoids macro conflict with <sys/ioctl.h>
+ */
+#define SOCKCLOSE    _IOR('f', 200, int)
+
+__BEGIN_DECLS
+extern int posix_devctl(int fildes,
+                        int dcmd,
+                        void *__restrict dev_data_ptr,
+                        size_t nbyte,
+                        int *__restrict dev_info_ptr);
+__END_DECLS
+
+#endif /* _POSIX_26_C_SOURCE == 200312L */
+#endif /* _DEVCTL_H */
diff --git a/posix/getconf.c b/posix/getconf.c
index 4ba9c0de..5ed0c005 100644
--- a/posix/getconf.c
+++ b/posix/getconf.c
@@ -405,6 +405,8 @@ static const struct conf vars[] =
     { "_POSIX_IPV6", _SC_IPV6, SYSCONF },
     { "_POSIX_RAW_SOCKETS", _SC_RAW_SOCKETS, SYSCONF },
 
+    { "_POSIX_26_VERSION", _SC_POSIX_26_VERSION, SYSCONF },
+
     { NULL, 0, SYSCONF }
   };
 
diff --git a/posix/posix_devctl.c b/posix/posix_devctl.c
new file mode 100644
index 00000000..a002274b
--- /dev/null
+++ b/posix/posix_devctl.c
@@ -0,0 +1,31 @@
+/* Copyright (C) 2019 Wind River Systems, Inc.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <http://www.gnu.org/licenses/>. */
+
+#include <devctl.h>
+#include <errno.h>
+
+/* The posix_devctl method is defined by POSIX 1003.26-2003.
+   The interface can simply pass the arguments to the regular system ioctl. */
+
+int posix_devctl(int fildes,
+                 int dcmd,
+                 void *__restrict dev_data_ptr,
+                 size_t nbyte,
+                 int *__restrict dev_info_ptr)
+{
+    return ENOSYS;
+}
+stub_warning (posix_devctl)
diff --git a/posix/sysconf.c b/posix/sysconf.c
index c3bf3b78..6419f873 100644
--- a/posix/sysconf.c
+++ b/posix/sysconf.c
@@ -148,6 +148,13 @@ __sysconf (int name)
     case _SC_GETPW_R_SIZE_MAX:
       return NSS_BUFLEN_PASSWD;
 
+    case _SC_POSIX_26_VERSION:
+#ifdef _POSIX_26_C_SOURCE
+      return _POSIX_26_VERSION;
+#else
+      return -1;
+#endif
+
     case _SC_ARG_MAX:
     case _SC_CHILD_MAX:
     case _SC_CLK_TCK:
diff --git a/posix/unistd.h b/posix/unistd.h
index 3f4e4262..c1495bdf 100644
--- a/posix/unistd.h
+++ b/posix/unistd.h
@@ -85,6 +85,10 @@ __BEGIN_DECLS
    creation of locales with the localedef utility.  */
 #define _POSIX2_LOCALEDEF       __POSIX2_THIS_VERSION
 
+/* If defined, the implementation supports the
+   implement posix_devctl() using ioctl()  */
+#define _POSIX_26_C_SOURCE 200312L
+
 /* X/Open version number to which the library conforms.  It is selectable.  */
 #ifdef __USE_XOPEN2K8
 # define _XOPEN_VERSION	700
diff --git a/sysdeps/mach/hurd/i386/libc.abilist b/sysdeps/mach/hurd/i386/libc.abilist
index 40f35ee0..cee2e591 100644
--- a/sysdeps/mach/hurd/i386/libc.abilist
+++ b/sysdeps/mach/hurd/i386/libc.abilist
@@ -2117,6 +2117,7 @@ GLIBC_2.27 wcstof64x F
 GLIBC_2.27 wcstof64x_l F
 GLIBC_2.28 fcntl64 F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.29 _hurd_port_move F
 GLIBC_2.29 posix_spawn_file_actions_addchdir_np F
diff --git a/sysdeps/unix/sysv/linux/aarch64/libc.abilist b/sysdeps/unix/sysv/linux/aarch64/libc.abilist
index 38db77e4..ca81a42f 100644
--- a/sysdeps/unix/sysv/linux/aarch64/libc.abilist
+++ b/sysdeps/unix/sysv/linux/aarch64/libc.abilist
@@ -2353,6 +2353,7 @@ GLIBC_2.28 mtx_timedlock F
 GLIBC_2.28 mtx_trylock F
 GLIBC_2.28 mtx_unlock F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_create F
 GLIBC_2.28 thrd_current F
diff --git a/sysdeps/unix/sysv/linux/alpha/libc.abilist b/sysdeps/unix/sysv/linux/alpha/libc.abilist
index 637bfce9..4695dca2 100644
--- a/sysdeps/unix/sysv/linux/alpha/libc.abilist
+++ b/sysdeps/unix/sysv/linux/alpha/libc.abilist
@@ -2228,6 +2228,7 @@ GLIBC_2.28 mtx_timedlock F
 GLIBC_2.28 mtx_trylock F
 GLIBC_2.28 mtx_unlock F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_create F
 GLIBC_2.28 thrd_current F
diff --git a/sysdeps/unix/sysv/linux/arm/be/libc.abilist b/sysdeps/unix/sysv/linux/arm/be/libc.abilist
index 1d54f71b..9188519c 100644
--- a/sysdeps/unix/sysv/linux/arm/be/libc.abilist
+++ b/sysdeps/unix/sysv/linux/arm/be/libc.abilist
@@ -139,6 +139,7 @@ GLIBC_2.28 mtx_timedlock F
 GLIBC_2.28 mtx_trylock F
 GLIBC_2.28 mtx_unlock F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_create F
 GLIBC_2.28 thrd_current F
diff --git a/sysdeps/unix/sysv/linux/arm/le/libc.abilist b/sysdeps/unix/sysv/linux/arm/le/libc.abilist
index ff7e8bc4..51074ab1 100644
--- a/sysdeps/unix/sysv/linux/arm/le/libc.abilist
+++ b/sysdeps/unix/sysv/linux/arm/le/libc.abilist
@@ -139,6 +139,7 @@ GLIBC_2.28 mtx_timedlock F
 GLIBC_2.28 mtx_trylock F
 GLIBC_2.28 mtx_unlock F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_create F
 GLIBC_2.28 thrd_current F
diff --git a/sysdeps/unix/sysv/linux/csky/libc.abilist b/sysdeps/unix/sysv/linux/csky/libc.abilist
index c3ed6546..532e70fb 100644
--- a/sysdeps/unix/sysv/linux/csky/libc.abilist
+++ b/sysdeps/unix/sysv/linux/csky/libc.abilist
@@ -1729,6 +1729,7 @@ GLIBC_2.29 remque F
 GLIBC_2.29 rename F
 GLIBC_2.29 renameat F
 GLIBC_2.29 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.29 revoke F
 GLIBC_2.29 rewind F
 GLIBC_2.29 rewinddir F
diff --git a/sysdeps/unix/sysv/linux/hppa/libc.abilist b/sysdeps/unix/sysv/linux/hppa/libc.abilist
index 99147538..bcb9a30c 100644
--- a/sysdeps/unix/sysv/linux/hppa/libc.abilist
+++ b/sysdeps/unix/sysv/linux/hppa/libc.abilist
@@ -2060,6 +2060,7 @@ GLIBC_2.28 mtx_timedlock F
 GLIBC_2.28 mtx_trylock F
 GLIBC_2.28 mtx_unlock F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_create F
 GLIBC_2.28 thrd_current F
diff --git a/sysdeps/unix/sysv/linux/i386/libc.abilist b/sysdeps/unix/sysv/linux/i386/libc.abilist
index 4fedf775..b8a5b719 100644
--- a/sysdeps/unix/sysv/linux/i386/libc.abilist
+++ b/sysdeps/unix/sysv/linux/i386/libc.abilist
@@ -2241,6 +2241,7 @@ GLIBC_2.28 mtx_timedlock F
 GLIBC_2.28 mtx_trylock F
 GLIBC_2.28 mtx_unlock F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_create F
 GLIBC_2.28 thrd_current F
diff --git a/sysdeps/unix/sysv/linux/m68k/coldfire/libc.abilist b/sysdeps/unix/sysv/linux/m68k/coldfire/libc.abilist
index 142595eb..faa89c50 100644
--- a/sysdeps/unix/sysv/linux/m68k/coldfire/libc.abilist
+++ b/sysdeps/unix/sysv/linux/m68k/coldfire/libc.abilist
@@ -140,6 +140,7 @@ GLIBC_2.28 mtx_timedlock F
 GLIBC_2.28 mtx_trylock F
 GLIBC_2.28 mtx_unlock F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_create F
 GLIBC_2.28 thrd_current F
diff --git a/sysdeps/unix/sysv/linux/m68k/m680x0/libc.abilist b/sysdeps/unix/sysv/linux/m68k/m680x0/libc.abilist
index 85e7746c..0f466224 100644
--- a/sysdeps/unix/sysv/linux/m68k/m680x0/libc.abilist
+++ b/sysdeps/unix/sysv/linux/m68k/m680x0/libc.abilist
@@ -2184,6 +2184,7 @@ GLIBC_2.28 mtx_timedlock F
 GLIBC_2.28 mtx_trylock F
 GLIBC_2.28 mtx_unlock F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_create F
 GLIBC_2.28 thrd_current F
diff --git a/sysdeps/unix/sysv/linux/microblaze/be/libc.abilist b/sysdeps/unix/sysv/linux/microblaze/be/libc.abilist
index 91dc1b83..c94215af 100644
--- a/sysdeps/unix/sysv/linux/microblaze/be/libc.abilist
+++ b/sysdeps/unix/sysv/linux/microblaze/be/libc.abilist
@@ -2344,6 +2344,7 @@ GLIBC_2.28 mtx_timedlock F
 GLIBC_2.28 mtx_trylock F
 GLIBC_2.28 mtx_unlock F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_create F
 GLIBC_2.28 thrd_current F
diff --git a/sysdeps/unix/sysv/linux/microblaze/le/libc.abilist b/sysdeps/unix/sysv/linux/microblaze/le/libc.abilist
index 3440e90f..3800da25 100644
--- a/sysdeps/unix/sysv/linux/microblaze/le/libc.abilist
+++ b/sysdeps/unix/sysv/linux/microblaze/le/libc.abilist
@@ -2344,6 +2344,7 @@ GLIBC_2.28 mtx_timedlock F
 GLIBC_2.28 mtx_trylock F
 GLIBC_2.28 mtx_unlock F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_create F
 GLIBC_2.28 thrd_current F
diff --git a/sysdeps/unix/sysv/linux/mips/mips32/fpu/libc.abilist b/sysdeps/unix/sysv/linux/mips/mips32/fpu/libc.abilist
index 5ee7b8c5..ffa2b8ac 100644
--- a/sysdeps/unix/sysv/linux/mips/mips32/fpu/libc.abilist
+++ b/sysdeps/unix/sysv/linux/mips/mips32/fpu/libc.abilist
@@ -2154,6 +2154,7 @@ GLIBC_2.28 mtx_timedlock F
 GLIBC_2.28 mtx_trylock F
 GLIBC_2.28 mtx_unlock F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_create F
 GLIBC_2.28 thrd_current F
diff --git a/sysdeps/unix/sysv/linux/mips/mips32/nofpu/libc.abilist b/sysdeps/unix/sysv/linux/mips/mips32/nofpu/libc.abilist
index 6cb6328e..1364a55e 100644
--- a/sysdeps/unix/sysv/linux/mips/mips32/nofpu/libc.abilist
+++ b/sysdeps/unix/sysv/linux/mips/mips32/nofpu/libc.abilist
@@ -2152,6 +2152,7 @@ GLIBC_2.28 mtx_timedlock F
 GLIBC_2.28 mtx_trylock F
 GLIBC_2.28 mtx_unlock F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_create F
 GLIBC_2.28 thrd_current F
diff --git a/sysdeps/unix/sysv/linux/mips/mips64/n32/libc.abilist b/sysdeps/unix/sysv/linux/mips/mips64/n32/libc.abilist
index ae7474c0..445628df 100644
--- a/sysdeps/unix/sysv/linux/mips/mips64/n32/libc.abilist
+++ b/sysdeps/unix/sysv/linux/mips/mips64/n32/libc.abilist
@@ -2160,6 +2160,7 @@ GLIBC_2.28 mtx_timedlock F
 GLIBC_2.28 mtx_trylock F
 GLIBC_2.28 mtx_unlock F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_create F
 GLIBC_2.28 thrd_current F
diff --git a/sysdeps/unix/sysv/linux/mips/mips64/n64/libc.abilist b/sysdeps/unix/sysv/linux/mips/mips64/n64/libc.abilist
index cdf040de..43674d74 100644
--- a/sysdeps/unix/sysv/linux/mips/mips64/n64/libc.abilist
+++ b/sysdeps/unix/sysv/linux/mips/mips64/n64/libc.abilist
@@ -2155,6 +2155,7 @@ GLIBC_2.28 mtx_timedlock F
 GLIBC_2.28 mtx_trylock F
 GLIBC_2.28 mtx_unlock F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_create F
 GLIBC_2.28 thrd_current F
diff --git a/sysdeps/unix/sysv/linux/posix_devctl.c b/sysdeps/unix/sysv/linux/posix_devctl.c
new file mode 100644
index 00000000..2fd7ae61
--- /dev/null
+++ b/sysdeps/unix/sysv/linux/posix_devctl.c
@@ -0,0 +1,116 @@
+/* Define the posix_devctl.c function.
+   Copyright (C) 2014-2019 Wind River Systems, Inc.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <http://www.gnu.org/licenses/>. */
+
+#include <devctl.h>
+#include <errno.h>
+
+/* The posix_devctl method is defined by POSIX 1003.26-2003.
+   The interface can simply pass the arguments to the regular system ioctl. */
+
+int posix_devctl(int fildes,
+                 int dcmd,
+                 void *__restrict dev_data_ptr,
+                 size_t nbyte,
+                 int *__restrict dev_info_ptr)
+{
+  /* POSIX 1003.26-2003 indicates that there is a "general goal
+   * of being able to implement posix_devctl() using the existing
+   * ioctl() interfaces provided in current UNIX systems and other
+   * POSIX implementations because the standard allows but does not
+   * require checking the size of the device data."
+   *
+   * As such this function will be implemented primarily using the
+   * existing ioctl() interface as permitted by the specification.
+   *
+   * Additionally the FACE Technical Standard 3.0, Section 3.2.1.2,
+   * requiret that FIONBIO and SOCKCLOSE command be supported.
+   */
+
+  int ret = 0;
+  __set_errno (0);
+
+  if (fildes < 0)
+  {
+    __set_errno (EBADF);
+    ret = -1;
+    goto on_error;
+  }
+
+  if (nbyte > 0 && dev_data_ptr == NULL)
+  {
+    __set_errno (EINVAL);
+    ret = -1;
+    goto on_error;
+  }
+
+  /* An OSS UoC shall provide:
+   * a. The SOCKCLOSE command in the posix_devctl() API
+   * b. The FIONBIO command in the posix_devctl() API
+   */
+  switch (dcmd)
+    {
+    case SOCKCLOSE:
+      if (nbyte > 0 || dev_data_ptr != NULL)
+      {
+        __set_errno (EINVAL);
+        ret = -1;
+        goto on_error;
+      }
+      ret = close(fildes);
+      break;
+    // FIONBIO is implemented in kernel
+    case FIONBIO:
+      if ((nbyte > 0 && nbyte < sizeof(int)) ||
+          dev_data_ptr == NULL)
+      {
+        __set_errno (EINVAL);
+        ret = -1;
+        goto on_error;
+      }
+    default:
+      ret = __ioctl(fildes, dcmd, dev_data_ptr);
+    }
+
+
+on_error:
+  /* ioctl() and posix_devctl() have different return value semantics.
+   *
+   * (from the Linux man-pages project)
+   * ioctl() will usually return success with zero, but a few ioctl() request use
+   * the return value as an output parameter and return a nonnegative value on
+   * success.  On error, -1 is returned, and errno is set appropriately.
+   *
+   * post_devctl(), upon successful completion, shall return zero; otherwise an
+   * error number shall be returned to indicate the error. The value returned via
+   * the dev_info_ptr argument is driver dependent.
+   *
+   * Based on this it can be assumed an ioctl() return value of -1 indicates
+   * an error, and the error number will be set in errno.
+   *
+   * Any nonnegative value returned is then considered to indicate success,
+   * and assumed to be an output parameter from ioctl().  Output parameters are
+   * transferred into dev_info_ptr, unless it is NULL.
+   */
+  if (dev_info_ptr != NULL) {
+      *dev_info_ptr = ret;
+  }
+
+  if (ret < 0)
+    ret = errno;
+
+  return ret;
+}
diff --git a/sysdeps/unix/sysv/linux/powerpc/powerpc32/fpu/libc.abilist b/sysdeps/unix/sysv/linux/powerpc/powerpc32/fpu/libc.abilist
index 7937f94c..be66b8c4 100644
--- a/sysdeps/unix/sysv/linux/powerpc/powerpc32/fpu/libc.abilist
+++ b/sysdeps/unix/sysv/linux/powerpc/powerpc32/fpu/libc.abilist
@@ -2188,6 +2188,7 @@ GLIBC_2.28 mtx_timedlock F
 GLIBC_2.28 mtx_trylock F
 GLIBC_2.28 mtx_unlock F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_create F
 GLIBC_2.28 thrd_current F
diff --git a/sysdeps/unix/sysv/linux/powerpc/powerpc32/nofpu/libc.abilist b/sysdeps/unix/sysv/linux/powerpc/powerpc32/nofpu/libc.abilist
index d6e35f31..08052232 100644
--- a/sysdeps/unix/sysv/linux/powerpc/powerpc32/nofpu/libc.abilist
+++ b/sysdeps/unix/sysv/linux/powerpc/powerpc32/nofpu/libc.abilist
@@ -2192,6 +2192,7 @@ GLIBC_2.28 mtx_timedlock F
 GLIBC_2.28 mtx_trylock F
 GLIBC_2.28 mtx_unlock F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_create F
 GLIBC_2.28 thrd_current F
diff --git a/sysdeps/unix/sysv/linux/powerpc/powerpc64/be/libc.abilist b/sysdeps/unix/sysv/linux/powerpc/powerpc64/be/libc.abilist
index 2268d689..29cc1e49 100644
--- a/sysdeps/unix/sysv/linux/powerpc/powerpc64/be/libc.abilist
+++ b/sysdeps/unix/sysv/linux/powerpc/powerpc64/be/libc.abilist
@@ -139,6 +139,7 @@ GLIBC_2.28 mtx_timedlock F
 GLIBC_2.28 mtx_trylock F
 GLIBC_2.28 mtx_unlock F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_create F
 GLIBC_2.28 thrd_current F
diff --git a/sysdeps/unix/sysv/linux/powerpc/powerpc64/le/libc.abilist b/sysdeps/unix/sysv/linux/powerpc/powerpc64/le/libc.abilist
index 7f61b14b..d734ad5e 100644
--- a/sysdeps/unix/sysv/linux/powerpc/powerpc64/le/libc.abilist
+++ b/sysdeps/unix/sysv/linux/powerpc/powerpc64/le/libc.abilist
@@ -2443,6 +2443,7 @@ GLIBC_2.28 mtx_timedlock F
 GLIBC_2.28 mtx_trylock F
 GLIBC_2.28 mtx_unlock F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_create F
 GLIBC_2.28 thrd_current F
diff --git a/sysdeps/unix/sysv/linux/riscv/rv64/libc.abilist b/sysdeps/unix/sysv/linux/riscv/rv64/libc.abilist
index 8935becc..8ee00d1b 100644
--- a/sysdeps/unix/sysv/linux/riscv/rv64/libc.abilist
+++ b/sysdeps/unix/sysv/linux/riscv/rv64/libc.abilist
@@ -2316,6 +2316,7 @@ GLIBC_2.28 mtx_timedlock F
 GLIBC_2.28 mtx_trylock F
 GLIBC_2.28 mtx_unlock F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_create F
 GLIBC_2.28 thrd_current F
diff --git a/sysdeps/unix/sysv/linux/s390/s390-32/libc.abilist b/sysdeps/unix/sysv/linux/s390/s390-32/libc.abilist
index e69dc7cc..25924ccb 100644
--- a/sysdeps/unix/sysv/linux/s390/s390-32/libc.abilist
+++ b/sysdeps/unix/sysv/linux/s390/s390-32/libc.abilist
@@ -2197,6 +2197,7 @@ GLIBC_2.28 mtx_timedlock F
 GLIBC_2.28 mtx_trylock F
 GLIBC_2.28 mtx_unlock F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_create F
 GLIBC_2.28 thrd_current F
diff --git a/sysdeps/unix/sysv/linux/s390/s390-64/libc.abilist b/sysdeps/unix/sysv/linux/s390/s390-64/libc.abilist
index 7d860001..6f44a2c1 100644
--- a/sysdeps/unix/sysv/linux/s390/s390-64/libc.abilist
+++ b/sysdeps/unix/sysv/linux/s390/s390-64/libc.abilist
@@ -2081,6 +2081,7 @@ GLIBC_2.28 mtx_timedlock F
 GLIBC_2.28 mtx_trylock F
 GLIBC_2.28 mtx_unlock F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_create F
 GLIBC_2.28 thrd_current F
diff --git a/sysdeps/unix/sysv/linux/sh/be/libc.abilist b/sysdeps/unix/sysv/linux/sh/be/libc.abilist
index fcb81618..c94d50a0 100644
--- a/sysdeps/unix/sysv/linux/sh/be/libc.abilist
+++ b/sysdeps/unix/sysv/linux/sh/be/libc.abilist
@@ -2064,6 +2064,7 @@ GLIBC_2.28 mtx_timedlock F
 GLIBC_2.28 mtx_trylock F
 GLIBC_2.28 mtx_unlock F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_create F
 GLIBC_2.28 thrd_current F
diff --git a/sysdeps/unix/sysv/linux/sh/le/libc.abilist b/sysdeps/unix/sysv/linux/sh/le/libc.abilist
index 3fd078d1..614baf09 100644
--- a/sysdeps/unix/sysv/linux/sh/le/libc.abilist
+++ b/sysdeps/unix/sysv/linux/sh/le/libc.abilist
@@ -2064,6 +2064,7 @@ GLIBC_2.28 mtx_timedlock F
 GLIBC_2.28 mtx_trylock F
 GLIBC_2.28 mtx_unlock F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_create F
 GLIBC_2.28 thrd_current F
diff --git a/sysdeps/unix/sysv/linux/sparc/sparc32/libc.abilist b/sysdeps/unix/sysv/linux/sparc/sparc32/libc.abilist
index 1ce1fe9d..f7ac53e1 100644
--- a/sysdeps/unix/sysv/linux/sparc/sparc32/libc.abilist
+++ b/sysdeps/unix/sysv/linux/sparc/sparc32/libc.abilist
@@ -2191,6 +2191,7 @@ GLIBC_2.28 mtx_timedlock F
 GLIBC_2.28 mtx_trylock F
 GLIBC_2.28 mtx_unlock F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_create F
 GLIBC_2.28 thrd_current F
diff --git a/sysdeps/unix/sysv/linux/sparc/sparc64/libc.abilist b/sysdeps/unix/sysv/linux/sparc/sparc64/libc.abilist
index 07507b86..99b1db4a 100644
--- a/sysdeps/unix/sysv/linux/sparc/sparc64/libc.abilist
+++ b/sysdeps/unix/sysv/linux/sparc/sparc64/libc.abilist
@@ -2117,6 +2117,7 @@ GLIBC_2.28 mtx_timedlock F
 GLIBC_2.28 mtx_trylock F
 GLIBC_2.28 mtx_unlock F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_create F
 GLIBC_2.28 thrd_current F
diff --git a/sysdeps/unix/sysv/linux/x86_64/64/libc.abilist b/sysdeps/unix/sysv/linux/x86_64/64/libc.abilist
index 5acf49db..d64b952c 100644
--- a/sysdeps/unix/sysv/linux/x86_64/64/libc.abilist
+++ b/sysdeps/unix/sysv/linux/x86_64/64/libc.abilist
@@ -2070,6 +2070,7 @@ GLIBC_2.28 mtx_timedlock F
 GLIBC_2.28 mtx_trylock F
 GLIBC_2.28 mtx_unlock F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_create F
 GLIBC_2.28 thrd_current F
diff --git a/sysdeps/unix/sysv/linux/x86_64/x32/libc.abilist b/sysdeps/unix/sysv/linux/x86_64/x32/libc.abilist
index 02d1bb97..3f0c5496 100644
--- a/sysdeps/unix/sysv/linux/x86_64/x32/libc.abilist
+++ b/sysdeps/unix/sysv/linux/x86_64/x32/libc.abilist
@@ -2366,6 +2366,7 @@ GLIBC_2.28 mtx_timedlock F
 GLIBC_2.28 mtx_trylock F
 GLIBC_2.28 mtx_unlock F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_create F
 GLIBC_2.28 thrd_current F
-- 
2.39.2

